torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp

```diff
@@ -0,0 +1,60 @@
+#include <torch/csrc/distributed/autograd/context/DistAutogradContainer.h>
+#include <c10/util/Exception.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+constexpr auto kContextIdBits = 48;
+
+DistAutogradContainer::DistAutogradContainer() : initialized_(false) {}
+
+DistAutogradContainer& DistAutogradContainer::init(int16_t worker_id) {
+  auto& container = getInstance();
+  container.worker_id_ = worker_id;
+  container.current_context_id_ = static_cast<int64_t>(worker_id)
+      << kContextIdBits;
+  container.initialized_ = true;
+  return container;
+}
+
+DistAutogradContainer& DistAutogradContainer::getInstance() {
+  static DistAutogradContainer container;
+  return container;
+}
+
+const DistAutogradContext& DistAutogradContainer::newContext() {
+  if (!initialized_) {
+    throw std::runtime_error(
+        "Need to initialize distributed autograd using "
+        "torch.distributed.autograd.init()");
+  }
+
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  autograd_context_.emplace(
+      current_context_id_, DistAutogradContext(current_context_id_));
+  return autograd_context_.at(current_context_id_++);
```

Comment:
nit: Let's check for overflow into the worker_id portion. It's unlikely to happen but if it does, the check will save you a lot of heartache.